### ISSUE TITLE
[bug] `short_paths` option leads to creation redundant subdirectories in C:/.conan after `conan remove`

### ISSUE THREAD

**Issue by Artalus** [conan-io/conan#7983]
I tested this with global override `CONAN_USE_ALWAYS_SHORT_PATHS=1`, but I trust it's just the same for every package with `short_paths=True` in conanfile.

### Environment Details (include every applicable attribute)
  * Windows:
  * VS 2013 (should be independent from compiler):
  * Conan 1.27.1 / 1.31:
  * Python 3.7.4 / 3.8:

### Steps to reproduce (Include if Applicable)
- Run `conan install` on conanfile.txt containing N packages with `short_paths` feature enabled
- Check `C:/.conan` (or your `CONAN_USER_HOME_SHORT` if set) for the number of folders (4×N)
- Run `conan remove *` and then `conan install` again
- Watch number of folders in shortener become 7×N
- Each repeated `remove+install` cycle will increase the number of folders

### Logs (Executed commands with output) (Include/Attach if Applicable)
```
C:\conantest
λ set CONAN_USE_ALWAYS_SHORT_PATHS=1
λ set CONAN_USER_HOME=C:\conantest
λ conan install -if b conanfile-win32.txt -pr conanprofile.vc120 -s build_type=Release
...
λ ls .conan/data | wc -l
38

bash-3.1$ ls /c/.conan | wc -l
152
bash-3.1$ for x in *; do if [[ -f $x/real_path.txt ]]; then echo "$x: $(cat $x/real_path.txt)"; fi done | grep zlib | /usr/bin/sort -t':' -b +2
99a5b5: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\export_source
03f603: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\package\25ae31d511843d8b24249c76bfff96b70097a22f
505002: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\scm_source
140320: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\source

C:\conantest
λ conan remove * -f
λ ls .conan/data | wc -l
0

bash-3.1$ ls /c/.conan | wc -l
114
bash-3.1$ for x in *; do if [[ -f $x/real_path.txt ]]; then echo "$x: $(cat $x/real_path.txt)"; fi done | grep zlib | /usr/bin/sort -t':' -b +2
99a5b5: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\export_source
505002: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\scm_source
140320: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\source

C:\conantest
λ conan install -if b conanfile-win32.txt -pr conanprofile.vc120 -s build_type=Release
...
λ ls .conan/data | wc -l
38

bash-3.1$ ls /c/.conan | wc -l
266
bash-3.1$ for x in *; do if [[ -f $x/real_path.txt ]]; then echo "$x: $(cat $x/real_path.txt)"; fi done | grep zlib | /usr/bin/sort -t':' -b +2
99a5b5: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\export_source
99a5b5a: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\export_source
03f603: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\package\25ae31d511843d8b24249c76bfff96b70097a22f
5050029: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\scm_source
505002: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\scm_source
140320: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\source
140320c: C:\conantest\.conan\data\zlib\1.2.8\trassir\sdk120\source
```
The quirk, I believe, is `util/windows.py : hashed_redirect`:
```python
    for length in range(min_length, max_length):
        redirect = os.path.join(base, full_hash[:length])
        if not os.path.exists(redirect): # <<<
            return redirect
```
, so either the check should be smth like `if not exists(redirect) or content(join(redirect, CONAN_REAL_PATH)) == join(base, path)`, or at least `conan remove` should also remove those `\source` subdirs. However the same thing happens for `package` shorten subdir, if instead of `conan remove` you do simply `rm -rf /c/conantest/.conan/data`, so I believe the enhanced check is a way to go.
Otherwise after a dozen of such remove+install cycles the hasher will run into `max_length` and resolve back to nondeterministic paths (which contradicts #3971 where it first apperaed).

Note of interest: in Conan 1.27.1 (which we currently use in our CI) the `source` subdirs are not created immediately - some of them (like source) appear immediately after `conan install` command, but some (like export_source) - only after `conan remove`. Which is kinda weird, but probably isn't a problem anyway since it's 4 versions behind.

**Comment by memsharded** [comment]
Actually, the problem happens with ``use_always_short_paths``, but apparently with ``short_paths=True`` in recipe it doesn't happen.

**Comment by Artalus** [comment]
That sounds slightly weird for me, because:
```python
def path_shortener(path, short_paths):
    use_always_short_paths = get_env("CONAN_USE_ALWAYS_SHORT_PATHS", False)
    short_paths = use_always_short_paths or short_paths # <<<
```
, and `short_paths` in the method argument seems to be arriving from the `ConanFileReference` instance.

--8<--

Meanwhile, simply adding check for `real_path.txt` is not a silver bullet, unfortunately. With this change https://github.com/trassir/conan/commit/3730975f617a36f9c87d5e5cf17fdc4fb74eee1c if I leave `C:/.conan` untouched but completely remove `HOME/.conan`, next `conan install` fails with this:
```zlib/1.2.8@trassir/sdk120: Downloaded recipe revision eb46e75781cc612d161817cfaed6df15
Traceback (most recent call last):
  File "c:\python37\lib\site-packages\conans\client\command.py", line 2051, in run
    method(args[0][1:])
  File "c:\python37\lib\site-packages\conans\client\command.py", line 506, in install
    lockfile=args.lockfile)
  File "c:\python37\lib\site-packages\conans\client\conan_api.py", line 94, in wrapper
    return f(api, *args, **kwargs)
  File "c:\python37\lib\site-packages\conans\client\conan_api.py", line 581, in install
    recorder=recorder)
  File "c:\python37\lib\site-packages\conans\client\manager.py", line 56, in deps_install
    False, update, remotes, recorder)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 118, in load_graph
    recorder, apply_build_requires=apply_build_requires)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 245, in _resolve_graph
    graph_lock=graph_lock)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 355, in _load_graph
    apply_build_requires=apply_build_requires)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 286, in _recurse_build_requires
    self._binary_analyzer.evaluate_graph(graph, build_mode, update, remotes, nodes_subset, root)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_binaries.py", line 343, in evaluate_graph
    self._evaluate_node(node, build_mode, update, remotes)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_binaries.py", line 167, in _evaluate_node
    self._process_node(node, pref, build_mode, update, remotes)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_binaries.py", line 221, in _process_node
    package_folder)
  File "c:\python37\lib\site-packages\conans\client\graph\graph_binaries.py", line 97, in _evaluate_cache_pkg
    assert node.prev, "PREV for %s is None: %s" % (str(pref), metadata.dumps())
AssertionError: PREV for OpenSSL/1.0.2d@trassir/sdk120:25ae31d511843d8b24249c76bfff96b70097a22f is None:
{
  "recipe": {
    "revision": "4d99e25f3577dc8eacb8dde46257a90e",
    "remote": "local-conan",
    "properties": {},
    "checksums": {
      "conanmanifest.txt": {
        "md5": "52b1d6758dc8b95a9cbb634c59c83b32",
        "sha1": "1a2111335ecb5e4cc99cf4dfce125c180493eb2e"
      },
      "conanfile.py": {
        "md5": "ba81c1409f06b7936a9cd6d54d59cb40",
        "sha1": "dc4be4de399cc14540908553d0b90e8f23202132"
      }
    }
  },
  "packages": {
    "25ae31d511843d8b24249c76bfff96b70097a22f": {
      "revision": null,
      "recipe_revision": null,
      "remote": null,
      "properties": {},
      "checksums": {}
    }
  }
}
```

**Comment by memsharded** [comment]
Short_paths is a complicated feature, and indeed it seems there is a bug in the removal of short_paths. I am proposing a fix in https://github.com/conan-io/conan/pull/7986, let's see how it goes.

Thanks very much for producing such a detailed report, it was really useful, we have suspected for a long time that there was some leaking in the short_paths folder, this really helped.

**Comment by Artalus** [comment]
7986 seems to be removing subdirs from short_paths home OK. To keep problems separated I've created #7995, and suppose this issue can be closed once the PR is successfully merged.

**Comment by memsharded** [comment]
Fixed in https://github.com/conan-io/conan/pull/7986, this will be released in 1.32
